File: src/sim/Dvfs.py
 
from m5.SimObject import SimObject
from m5.params import *

# This was a helpful guide on how to call C++ from python in gem5:
# https://www.mail-archive.com/gem5-users@gem5.org/msg09060.html
# In particular, don't forget that:
# - the creation of the C++ objects is decoupled in time from the python ones,
#   as C++ objects are created in bulk in m5.instantiate(). Therefore, before
#   calling m5.instantiate() trying to call a method on a python object doesn't
#   call the corresponding C++ object, it returns an error!
# - if the python object is not registered as child of root, its C++ object will
#   never be created, so you need to do
#   dvfs = Dvfs(fuffa = True)
#   root.add_child('Dvfs',dvfs)
#   before m5.instantiate() or it will never work!

class Dvfs(SimObject):
    type = 'Dvfs'
    cxx_class = 'Dvfs'
    abstract = False
    cxx_header = "sim/dvfs.hh"
    
    recreateFiles = Param.Bool("If true, do not append to stats.txt and ruby.stats")
    
    @classmethod
    def export_methods(cls,code):
        code('''
int getUpdateCounter();
int getNumRouters();
int getNumCpus();
double getTotalRouterPower(int i);
double getStaticRouterPower(int i);
double getDynamicRouterPower(int i);
double getClockRouterPower(int i);
double getAverageRouterFrequency(int i);
double getAverageCpuFrequency(int i);
double getAverageRouterVoltage(int i);
double getAverageCpuVoltage(int i);
void setCpuTemperature(int i, double t);
void setRouterTemperature(int i, double t);
void runThermalPolicy();
''')

File: src/sim/dvfs.cc
 
#include "dvfs.hh"
#include <cassert>

using namespace std;

Dvfs::Dvfs(DvfsParams *p) : SimObject(p), recreateFiles(p->recreateFiles)
{
    assert(theInstance==0); //Don't call it twice
    theInstance=this;
}

Dvfs *Dvfs::instance()
{
    //assert(theInstance); //Has to exist
    return theInstance;
}

void Dvfs::configure(int numRouters)
{
    totalRouterPower.resize(numRouters,0);
    staticRouterPower.resize(numRouters,0);
    dynamicRouterPower.resize(numRouters,0);
    clockRouterPower.resize(numRouters,0);
    routerFrequency.resize(numRouters,0);
    routerVoltage.resize(numRouters,0);
    routerTemperature.resize(numRouters,0);
    updateCounter=0;
}

void Dvfs::configureCpu(int numCpus)
{
    cpuFrequency.resize(numCpus,0);
    cpuVoltage.resize(numCpus,0);
    cpuTemperature.resize(numCpus,0);
}

int   Dvfs::getUpdateCounter()                { return updateCounter; }
int   Dvfs::getNumRouters()                   { return totalRouterPower.size(); }
int   Dvfs::getNumCpus()                      { return cpuFrequency.size(); }
double Dvfs::getTotalRouterPower(int i)       { return totalRouterPower.at(i); }
double Dvfs::getStaticRouterPower(int i)      { return staticRouterPower.at(i); }
double Dvfs::getDynamicRouterPower(int i)     { return dynamicRouterPower.at(i); }
double Dvfs::getClockRouterPower(int i)       { return clockRouterPower.at(i); }
double Dvfs::getAverageRouterFrequency(int i) { return routerFrequency.at(i); }
double Dvfs::getAverageCpuFrequency(int i)    { return cpuFrequency.at(i); }
double Dvfs::getAverageRouterVoltage(int i)   { return routerVoltage.at(i); }
double Dvfs::getAverageCpuVoltage(int i)      { return cpuVoltage.at(i); }
void   Dvfs::setCpuTemperature(int i, double t)    { cpuTemperature.at(i)=t; }
void   Dvfs::setRouterTemperature(int i, double t) { routerTemperature.at(i)=t; }

void Dvfs::runThermalPolicy()
{
    if(thermalPolicyCallback) thermalPolicyCallback(this);
//    for(int i=0;i<cpuTemperature.size();i++)
//        cout<<"cpu"<<i<<" "<<cpuTemperature.at(i)<<endl;
//    for(int i=0;i<routerTemperature.size();i++)
//        cout<<"router"<<i<<" "<<routerTemperature.at(i)<<endl;
}

Dvfs::~Dvfs() {}


Dvfs *Dvfs::theInstance=0;

//

Dvfs *DvfsParams::create()
{
    return new Dvfs(this);
}
